packages/xeus/src/worker.ts

```diff
@@ -173,40 +192,271 @@ export abstract class EmpackedXeusRemoteKernel extends XeusRemoteKernelBase {
+        newInstalledPkgs[filename] = installedPkg;
+      }
+    });
+    return newInstalledPkgs;
+  }
+
+  protected getPackageName(specs: string) {
+    const nameMatch = specs.match(/^([a-zA-Z0-9_-]+)/);
+
+    if (!nameMatch) {
+      return null;
+    }
+
+    const packageName = nameMatch[1];
+    return packageName;
+  }
+
+  protected addSpecs(specs: string[], currentSpecs: { [key: string]: string }) {
+    const updatedCurrentSpecs = { ...currentSpecs };
+
+    specs.forEach((spec: string) => {
+      const pkgName = this.getPackageName(spec);
+      if (pkgName) {
+        updatedCurrentSpecs[pkgName] = spec;
+      }
+    });
+    return updatedCurrentSpecs;
+  }
+
+  protected deleteSpecs(
+    specs: string[],
+    currentSpecs: { [key: string]: string },
+    newInstalledPackagesMap: { [name: string]: ISolvedPackage },
+    updatedInstalled: ISolvedPackages,
+    type: string
+  ) {
+    const updatedCurrentSpecs = { ...currentSpecs };
+    let updatedInstalledPkgs = { ...updatedInstalled };
+
+    specs.forEach((spec: string) => {
+      const pkgName = this.getPackageName(spec);
+      if (pkgName && !newInstalledPackagesMap[pkgName]) {
+        this.logger.log(`Package ${pkgName} is not in the installed list`);
+      } else {
+        if (pkgName) {
+          // the special case when a user deletes pip
+          if (pkgName === 'pip') {
+            this.logger.log('Be aware of pip is being removed');
+          }
+          if (updatedCurrentSpecs[pkgName]) {
+            delete updatedCurrentSpecs[pkgName];
+            if (type === 'uninstall') {
+              updatedInstalledPkgs = this.deleteFromInstalledPackages(
+                pkgName,
+                updatedInstalledPkgs
+              );
+            }
+          } else {
+            updatedInstalledPkgs = this.deleteFromInstalledPackages(
+              pkgName,
+              updatedInstalledPkgs
+            );
+          }
+        }
+      }
+    });
+
+    return { specs: updatedCurrentSpecs, installed: updatedInstalledPkgs };
+  }
+
+  protected identifySpecs(
+    specs: string[],
+    currentSpecs: { [key: string]: string },
+    currentPipSpecs: { [key: string]: string }
+  ) {
+    const result = { fromCondaPkgs: false, fromPipPkgs: false };
+    specs.forEach((spec: string) => {
+      const pkgName = this.getPackageName(spec);
+      if (pkgName) {
+        if (currentSpecs[pkgName]) {
+          result.fromCondaPkgs = true;
+        } else if (currentPipSpecs[pkgName]) {
+          result.fromPipPkgs = true;
+        }
+      }
+    });
+    return result;
+  }
+
+  protected updateCurrentSpecs(
+    specs: string[],
+    pipSpecs: string[],
+    type: string
+  ): {
+    specs: { [key: string]: string };
+    pipSpecs: { [key: string]: string };
+    installed: ISolvedPackages;
+  } {
+    // copy data before making manupulations
+    let newSpecs = { ...this._currentSpecs };
+    let newPipSpecs = { ...this._currentPipSpecs };
```

Comment:
It looks to me we don't need and shouldn't apply the same logic of conda specs to pip specs.

Merging pip specs for each individual pip install will have the wrong byproduct of re-processing all pip installs that occurred before. I believe we shouldn't do this for pip packages, and keep the current logic. In the end, the current pip install logic works as intended, only the conda install was flawed. 

e.g. if one does:

```
%pip install ipywidgets  # works fine, this will print "successfully installed ipywidgets"

%pip install numpy  # works fine, but it will wrongfully report "ipywidgets already installed" which is not something we want. We don't want to process old pip specs
```

Am I missing something?

Reply:
I'm unresolving this, this is not resolved?

At least I understand that we're not doing anything with `this._currentPipSpecs` apart from assigning values to it. So we should simplify that code and get rid of any code interacting with `this._currentPipSpecs`